### PR TITLE
Hide application link for waypoint in graph

### PR DESCRIPTION
**Pull request by josunect**
**Describe the change**

It looks like the graph generation is not looking at the labels of the workload, it is looking at the [source_canonical_service](https://github.com/kiali/kiali/blob/master/graph/telemetry/istio/istio.go#L280) to fill the fill the app field. 

That's the reason to include the waypoint check in the frontend code. 

**Issue reference**

Ref. https://github.com/kiali/kiali/issues/7022

### CODE REVIEW

**Review by jshaughn — COMMENTED** [review]
The fundamental problem here is that we're trying to introduce Istio infrastructure into the application graph. To some degree we do this already with sidecar proxies, but because they are bundled with application pods it's not as explicit and obtrusive.    But with waypoints Istio is leaking this infra telemetry into the application namespaces. I'm not even sure Istio should be generating this telemetry, and maybe it will go away.  But for now it does exist and so I guess we need to deal with it.  We can't just treat these nodes like standard application nodes, and I think there may be multiple places where we will find ourselves wanting to do some special-case handling.

I don't really love this approach.  Instead, I'd recommend that we make this determination on the server side, as part of what happens in the new "ambient" appender.   We can add an `IsWaypoint` metadata and just pass it up as node data.  Then the UI can do what it wants to do whenever there is special handling involved.

**Comment by josunect** [comment]
> The fundamental problem here is that we're trying to introduce Istio infrastructure into the application graph. To some degree we do this already with sidecar proxies, but because they are bundled with application pods it's not as explicit and obtrusive. But with waypoints Istio is leaking this infra telemetry into the application namespaces. I'm not even sure Istio should be generating this telemetry, and maybe it will go away. But for now it does exist and so I guess we need to deal with it. We can't just treat these nodes like standard application nodes, and I think there may be multiple places where we will find ourselves wanting to do some special-case handling.
> 
> I don't really love this approach. Instead, I'd recommend that we make this determination on the server side, as part of what happens in the new "ambient" appender. We can add an `IsWaypoint` metadata and just pass it up as node data. Then the UI can do what it wants to do whenever there is special handling involved.

I like this approach, it can also be really useful for other issues.
In the last commit I've included the _isOutOfMesh_ to fix this issue: https://github.com/kiali/kiali/issues/7027 as I think it could also be part of this appender.

**Comment by jshaughn** [comment]
> I like this approach, it can also be really useful for other issues.
In the last commit I've included the isOutOfMesh to fix this issue: https://github.com/kiali/kiali/issues/7027 as I think it could also be part of this appender.

+1.  In general, I think it is best if the client-side is as dumb as possible, and worried specifically about "presentation", minimizing any other sort of logic.  With this approach we defer the logic to the back-end and the front-end just reacts to the information.

File: frontend/src/types/Graph.ts [hunk 1]
```diff
@@ -474,6 +474,8 @@ export interface DecoratedGraphNodeData extends GraphNodeData {
   httpOut: number;
   // true if has istio namespace
   isIstio?: boolean;
+  // true if is an Ambient Istio waypoint
+  isWaypoint?: boolean;
```
Comment (jshaughn) [inline frontend/src/types/Graph.ts:478]: This should actually be moved to the base type,`GraphNodeData`.    You may also want to add this in the `NodeAttr` definition.

File: graph/telemetry/istio/appender/ambient.go [hunk 2]
```diff
@@ -34,11 +34,13 @@ func (a AmbientAppender) AppendGraph(trafficMap graph.TrafficMap, globalInfo *gr
 	log.Trace("Running ambient appender")
 
 	if !a.Waypoints {
-		a.removeWaypointEntries(trafficMap, globalInfo, namespaceInfo)
+		a.isWaypoint(trafficMap, globalInfo, true)
+	} else {
+		a.isWaypoint(trafficMap, globalInfo, false)
 	}
```
Comment (jshaughn) [inline graph/telemetry/istio/appender/ambient.go:40]: Fine as is, but could reduce to just one line:

`a.isWaypoint(trafficMap, globalInfo, !a.Waypoints)`

**Review by jshaughn — COMMENTED** [review]
@josunect Looks good, this is the approach we talked about.  A couple of comments.  Also, if you wanted to you could add waypoints to graph find/hide, but maybe it's not necessary given that we already have the Display option.

File: frontend/src/types/Graph.ts (hunk 1 above)
Comment (josunect) [reply frontend/src/types/Graph.ts]: Moved and added also to NodeAttr

File: graph/telemetry/istio/appender/ambient.go (hunk 2 above)
Comment (josunect) [reply graph/telemetry/istio/appender/ambient.go]: oh! Good catch

File: frontend/src/pages/Graph/SummaryPanelNode.tsx [hunk 3]
```diff
@@ -224,11 +224,12 @@ export class SummaryPanelNodeComponent extends React.Component<SummaryPanelNodeC
...
-              <div className={nodeInfoStyle}>
-                {renderBadgedLink(nodeData)}
-                {renderHealth(nodeData.health)}
-              </div>
+              {!nodeData.isWaypoint && (
+                <div className={nodeInfoStyle}>
+                  {renderBadgedLink(nodeData)}
+                  {renderHealth(nodeData.health)}
+                </div>
+              )}
```
[the remark this answers is not in the document]
Comment (josunect) [reply frontend/src/pages/Graph/SummaryPanelNode.tsx:232]: But this is really the problem: 

image

The waypoint proxy doesn't have an app label, so the app link or text should not be there, and also the health (Probably is using the label to select the workload to get the health so it does not exist and appears as unknown)

[the remark this answers is not in the document]
Comment (hhovsepy) [reply frontend/src/pages/Graph/SummaryPanelNode.tsx:232]: @josunect @jshaughn wdyt to show some label or info near the '**Application**', like '(Waypoint proxy)' or something similar, as the application link or text does not exist anymore.

[the remark this answers is not in the document]
Comment (josunect) [reply frontend/src/pages/Graph/SummaryPanelNode.tsx:232]: Maybe we can add something like this? 

image

[the remark this answers is not in the document]
Comment (hhovsepy) [reply frontend/src/pages/Graph/SummaryPanelNode.tsx:232]: +1

[the remark this answers is not in the document]
Comment (jshaughn) [reply frontend/src/pages/Graph/SummaryPanelNode.tsx:233]: I'm also +1.  One suggested change, please update the PfBadge from `W` to `WP`.  I think it's a little confusing as `W` is the PfBadge for Workload, which is used a lot.  `WP` is already used for WasmPlugin, but that is much less of an overlap, and never shows up in the graph side panel.

File: frontend/src/types/Graph.ts [hunk 4]
```diff
@@ -399,6 +399,8 @@ export interface GraphNodeData {
   isOutside?: boolean;
   isRoot?: boolean;
   isServiceEntry?: SEInfo;
+  // true if is an Ambient Istio waypoint
```
Comment (jshaughn) [inline frontend/src/types/Graph.ts:402]:
```suggestion
```

File: graph/telemetry/istio/appender/ambient.go [hunk 5]
```diff
@@ -33,12 +33,10 @@ func (a AmbientAppender) AppendGraph(trafficMap graph.TrafficMap, globalInfo *gr
...
 	log.Trace("Running ambient appender")
 
-	if !a.Waypoints {
-		a.removeWaypointEntries(trafficMap, globalInfo, namespaceInfo)
-	}
+	a.isWaypoint(trafficMap, globalInfo, !a.Waypoints)
 }
 
-func (a AmbientAppender) removeWaypointEntries(trafficMap graph.TrafficMap, globalInfo *graph.AppenderGlobalInfo, namespaceInfo *graph.AppenderNamespaceInfo) {
+func (a AmbientAppender) isWaypoint(trafficMap graph.TrafficMap, globalInfo *graph.AppenderGlobalInfo, remove bool) {
```
Comment (jshaughn) [inline graph/telemetry/istio/appender/ambient.go:39]: Very minor, but `isWaypoint` seems like a function that would test something and return a bool.  Maybe rename to something like `processWaypoints` or `handleWaypoints` ?

**Review by jshaughn — COMMENTED** [review]
Looking good, some final comments...

File: graph/telemetry/istio/appender/ambient.go (hunk 5 above)
Comment (josunect) [reply graph/telemetry/istio/appender/ambient.go]: Changed to handleWaypoints

**Review by jshaughn — APPROVED** [review]
Perfecto!  With the work we are doing to handle it, I hope Istio continues to generate the waypoint telem :-)

**Review by hhovsepy — APPROVED** [review]
thanks @josunect looks good:
Screenshot from 2024-01-18 17-34-42

**Pull request merged**